### ISSUE TITLE
Blueprints Parsing strategy

### ISSUE THREAD

**Issue by adamziel** [WordPress/php-toolkit#13]
Let's find the right approach to parsing Blueprints in PHP.

The final API should enable the following operations (the syntax may change): 

```php
$blueprint = $parser
   ->parse('{ "wpVersion": "6.4", "steps": [ { "step": "installPlugin", "slug": "gutenberg" } ] }')
   ->getBlueprint();
// $blueprint->wpVersion is "6.4"

$errors = $parser
    ->parse('{ "wpVersion": "oops", "steps": [ { "step": "invalidStep" } ] }')
    ->getErrors();
// $errors should inform about invalid wpVersion and step
```

## Design goals

* **Extensibility** – At minimum, plugins should be able to register new steps and resource handlers. At maximum, they could register new top-level properties and extend the existing semantics.
* **Clarity** – Even with a few plugins, it should be easy to tell what is the current Blueprint definition and where each part comes from.
* **Static analysis** – Blueprints should have a JSON schema and, as a PHP library, IDE support for autocompletion.
* **Small footprint** – The parser should be as small as it can reasonably be, without involving dozens of MBs of libraries.

## Data flow

![Blueprints data flow](https://github.com/WordPress/blueprints/assets/205419/e7e583b1-474a-42f0-bf9d-15209a5f4298)

### Static definitions are convenient but less extensible

One way to define a Blueprint is with PHP classes:

```php
class Blueprint {
	/** @Type("string") */
	public $wpVersion;
	/** @Type("Step[]") */
	public $steps;
}

class UnzipStep extends Step {
	/** @Type("Resource") */
	public $file;
	/** @Type("string") */
	public $toPath;
}
```

From there, we could use PHP Reflections to validate the JSON input and match its keys with the related PHP properties. I [explored that approach using the Symfony Forms component](https://github.com/WordPress/blueprints/pull/12) and here's what I've learned

**Advantages:**

* Schema, code, and validation rules are all centralized
* Code Completion and static validation works really well 
* JSON Schema is reasonably easy to autogenerate from those PHP definitions
* Registering new steps, resource definitions, and even validators is easy since they're resolved at runtime
* Symfony Forms are really good for validating the JSON input, loading into PHP classes, and yielding developer-friendly error messages. We can also leverage their URLValidator and other niceties.

**Disadvantages:**

* Extending beyond new Steps and Resources is difficult. For example, to add a new top-level property you'd have to swap the `Blueprint` class. It could be supported via a service container, but only a single plugin could benefit from that. Alternatively, we could enable registering new properties via the PHP Reflection API. It feels really off, but perhaps it's fine?
* Symfony Forms are a complex 20MB library. We could, however, rebuild the relevant subset of those features in ~1000 lines of code.
* Complex type definitions require custom code on our end anyway. For example, implementing a top-level `$plugins` shorthand of type `Array<string | PluginStepDefinition>` isn't straightworward

### Dynamic definitions are inconvenient but infinitely extensible

Another way to define a Blueprint is with JSON schema:

```php
$blueprint_schema = [
	"wp" => [
		"type" => "string",
	],
	"steps" => [
		"type" => "array",
		"items" => {
			"$ref" => "#/definitions/StepDefinition"
		}
	],
	"StepDefinition" => [
		"type" => "object",
		"oneOf" => [
			"type" => "object",
			"additionalProperties"=> false,
			"properties" => [
				"file" => [
					"$ref": "#/definitions/Resource"
				],
				"toPath" => [
					"type" => "string"
				]
			]
		]
	]
];
```

From there, we may use JSON Schema validators to process the input. This is what the [TypeScript Blueprint compiler](https://github.com/WordPress/wordpress-playground/blob/trunk/packages/playground/blueprints/src/lib/compile.ts) does.

**Advantages:**

* The schema may be extended in any way imaginable, it's just an array
* No special libraries are required, other than a JSON Schema validator
* JSON Schema is the source of truth, it doesn't need to be autogenerated
* It's easy to express complex type definitions such as `Array<string | PluginStepDefinition>` 

**Disadvantages:**

* Schema and PHP handlers are separate concepts, discrepancies will lead to bugs and some will be difficult to identify. We'll need a top-notch test coverage.
* No code completion or static PHP analysis. Blueprints would be arrays with no associated static metadata.
* JSON schema validation produces cryptic errors, especially when discriminators are involved.
* There's no easy way to include custom semantics, like telling the PHP library "this field is a URL, please validate it as such".

## What does the best of both worlds look like?

Is it possible to benefit both from PHP static typing AND the type expresiveness of JSON schema without maintaining two distinct subsystems? 

Here's one idea:

```php
/** @JsonType("Blueprint") */
class Blueprint {
	/**
	@JsonSchema({
		"anyOf": [
			{ "$ref": "#/types/Resource" },
			{ "type": "string" }
		]
	})
	*/
	public $wpVersion;

	/**
	@JsonSchema({
		"type": "array",
		"items": {
			"$ref": "#/types/Step"
		}
	})
	*/
	public $steps;
}

/** @JsonType("Step") */
class UnzipStepDefinition extends StepDefinition {
	/** @JsonSchema({ "$ref": "#/types/Resource" }) */
	public $file;
	/** @JsonSchema({ "type": "string" }) */
	public $toPath;
}
```

In this scenario:

* JSON Schema would be generated based on the PHP classes
* Any discrepancies between the Schema and the class definition would trigger an error
* Validation would happen based on the JSON Schema with custom improvements to error reporting logic
* Validated JSON data would be mapped 1:1 to the PHP objects

[This library](https://api-platform.com/docs/core/json-schema/) could generate the code.

What about extensibility? Registering new steps and resources would require creating a new class annotated as `/** @JsonType("Step") */`.

Now, say the WooCommerce plugin wants to register new top-level property called "products". It could work as follows:

```php
class WooBlueprint extends Blueprint {
	/** @JsonSchema({ "type": "array", "items": "..." }) */
	public $products;
}

$blueprintSchema->extend(WooBlueprint::class)
// extend() modifies the schema AND creates a `Blueprint::$products` class property via the Reflections API

class ProductsPropertyHandler {
    public function process(Blueprint $blueprint) {
        /** @var WooBlueprint $blueprint */
        // Code completion works!
        foreach($blueprint->products as $product) {
            // ...
        }
    }
}
```

It's not perfect, as the `process()` method doesn't use the `WooBlueprint` type hint, but it is something.

cc @dmsnell @bgrgicak

**Comment by adamziel** [comment]
Another idea: Use JSON Schema as the source of truth and generate code for the PHP models and validators from that. Here's a library that seems quite capable: https://github.com/swaggest/php-code-builder.

[PR 14](https://github.com/WordPress/blueprints/pull/14) explores that approach. So far, I like this approach the most.

Here's another interesting project: https://typeschema.org/

**Comment by bgrgicak** [comment]
I like the idea of combining the two approaches. Building PHP from a schema sounds odd, but looking at your PR and Swaggest it seems promising. Let's keep going in that direction, we can always change our minds later.

**Comment by adamziel** [comment]
I'm not fully happy with the existing code generator – it creates `Model` classes for data storage and separate `Builder` classes that inherit from Models and perform serialization, deserialization, and validation. The culprit is `$blueprintBuilder->setPreferredVersions()` expects `BlueprintPreferredVersionsBuilder`, not just `BlueprintPreferredVersions`. I'd rather only have simple data storage classes unrelated to validation, serialization, etc.

The data pipeline looks as follows:

* Raw JSON is parsed to a data object.
* A Data object is mapped to a Blueprint model. This involves small transformations like mapping string URLs to `URLReference` instances.
* In addition, the developer may create a Blueprint model directly by using the API.
* The transformed Blueprint model is validated for things like not-null constraints.

There are two separate tasks here:

* Mapping JSON to model instances
* Validating model instances

I've done a few experiments to explore different paths forward:

* https://github.com/WordPress/blueprints/pull/17
* https://github.com/WordPress/blueprints/pull/19
* https://github.com/WordPress/blueprints/pull/20
* https://github.com/WordPress/blueprints/pull/21

It seems like we have the following ways forward for the JSON Schema -> PHP code approach:

* Expose Swaggest `Models`, hide `Builders` internally as "Validators" or "DataProcessors", rewrite between `Models` and `DataProcessors` every time we need to map or validate. Downside: It will get in our way all the time. It's like fighting the API, not working with it. 
* Use Swaggest only to generate simple plain `Model` classes, use `jsonmapper` for Mapping, use https://opis.io/ for validation. Downside: One library becomes three, we'll be more prone to errors and require more maintenance.
* Fork [the Jane PHP](https://github.com/WordPress/blueprints/pull/19) library to resolve the kinks. Tradeoff: We'll need many MBs of Symfony validation libraries.
* Write the models manually OR generate JSON Schema from PHP code. Tradeoff: We'll still need a way to map raw input to the model instances and validate them, and that goes back to the problems that came up in https://github.com/WordPress/blueprints/pull/12.

🤔

**Comment by adamziel** [comment]
> Expose Swaggest Models, hide Builders internally as "Validators" or "DataProcessors", rewrite between Models and DataProcessors every time we need to map or validate. Downside: It will get in our way all the time. It's like fighting the API, not working with it.

I'd like to explore that option. The Swaggest library solves mapping, validation, and serialization well. The only issue is that it also stores a bunch of not-so-useful data in the model objects. I don't like keeping garbage data in there and it also pollutes the `var_dump()` output, but perhaps it's not a bad trade-off to take given there's no perfect option.

**Comment by bgrgicak** [comment]
> Write the models manually OR generate JSON Schema from PHP code. 

If we do this could we explore more advanced options later?
I don't want tooling to block us and if v1 requires some manual work it's fine IMO. We can always automate later.
We just need to ensure that support for extending blueprints doesn't get broken. But extending could also be v2.

**Comment by adamziel** [comment]
In https://github.com/WordPress/blueprints/pull/19#issuecomment-1956210539 I just realized that we may need two stages of validation:

1. JSON validation to check if it conforms to the schema
2. Blueprint object validation to confirm it was constructed properly

Here's a situation where that would matter:

```php
$blueprint = Blueprints::fromJSON($jsonData);
$blueprint->addPlugin( ["invalid data – this is not supposed to be an array"] );
Blueprints::run($blueprint);
```

Here's the data flow as I imagine it:

![Blueprints data flow](https://github.com/WordPress/blueprints/assets/205419/e7e583b1-474a-42f0-bf9d-15209a5f4298)

**Comment by adamziel** [comment]
> I don't want tooling to block us and if v1 requires some manual work it's fine IMO. We can always automate later.

Yeah good point, I'm leaning more and more towards this option.

**Comment by bgrgicak** [comment]
> $blueprint->addPlugin( ["invalid data – this is not supposed to be an array"] );

Should we validate it? It's a great feature, but as a developer, if I make a mistake I'm used to my app crashing.

**Comment by adamziel** [comment]
> Should we validate it? It's a great feature, but as a developer, if I make a mistake I'm used to my app crashing.

Crashing is fine. I'm worried about the error message – confusing stack traces and messages will lead to confusing issues. Also if the library can produce an invalid Blueprint, someone will export it and use it on a website or in the Blueprints directory and it will eventually come back to this repo. Also, I'm sure I myself will make a bunch of mistakes when working with Blueprints and I'd love the library to prevent me from producing invalid Blueprints as much as possible.

**Comment by adamziel** [comment]
## PHP weak type system pushes us to recreate type checks in runtime

The problem here is just PHP, the language. TypeScript gives us data mapping and validation for free with its pattern matching interfaces. PHP, however, pushes us to build the following bits on our own:

![Blueprints – what TypeScript gives us for free](https://github.com/WordPress/blueprints/assets/205419/2ad333c8-bc96-4951-bb43-4feac439a4f4)

```ts
interface Blueprint {
    wpVersion: string;
    steps: StepDefinition[];
}

// TypeScript won't let us write an invalid Blueprint. This
// results in a type error:
const blueprintWrittenByHand: Blueprint = {
    wpVersion: {}
}; 

const rawBlueprint = JSON.parse(input);
if(!validateWithSchema(rawBlueprint, schema)) {
    throw new Error();
}

// Now we can just say `blueprint` is of type `Blueprint` without
// creating new objects or mapping any data between them:
const blueprint = rawBlueprint as Blueprint;

// From now on, we can be sure blueprint.steps is an array of StepDefinition objects

// Once the Blueprint adheres to the schema, TypeScript won't let us corrupt it:
blueprint.steps.push("A string"); // type error
blueprint.wpVersion = null; // type error
```

In PHP, however, the type system is basic and don't give us nearly the same guarantees:

```php
class Blueprint {
    public function __construct(
        string $wpVersion,
        array $steps
    ) {}
}

// This is legal:
$blueprintWrittenByHand = new Blueprint("6.3", ["I'm a string"]);

$rawBlueprint = json_decode($_POST['blueprint']);
if(!validateJson($rawBlueprint, $schema)) {
    throw new Exception();
}
// We cannot just set $rawBlueprint's type to Blueprint.
// Instead, we must painfully instantiate new objects and
// rewrite the entire deeply nested data structure.
$blueprint = mapRawDataToModels( $rawBlueprint );

// Even now, we must remain careful
array_push($blueprint->steps, 'string'); // Perfectly legal

// We can't even use readonly properties because they were introduced in 
// PHP 8 while this library aims to support PHP 7
```

The only somewhat type-safe model strategy in PHP seems to be this:

```php
class Blueprint {
    public function __construct(
        private string $wpVersion,
        private array $steps
    ) {
        foreach($steps as $step) {
            if(!($step instanceof StepDefinitionInterface)) {
                throw new InvalidArgumentException();
            }
        }
        // ...
    }
}
```

But that's a maintenance horror and it still requires mapping data from parsed JSON to class instances.

## Perhaps we could just add stronger typing?

If PHP gave us the same type errors as TypeScript, we could just reuse the object created by the original `json_decode` call. Perhaps the [Psalm](https://psalm.dev/) library could get us there for free? cc @dmsnell

**Comment by adamziel** [comment]
### PHP Psalm evaluation

[Psalm](https://psalm.dev/) is lenient with certain errors, but solves a few important typing and validation problems:

```php
<?php

/**
 * @psalm-type InstallPluginStep = array{
 *   step: "installPlugin",
 *   slug: string
 * }
 * @psalm-type RmStep = array{
 *   step: "rm",
 *   path: string
 * }
 * @psalm-type StepDefinition = RmStep|InstallPluginStep
 *
 * @psalm-type Blueprint = array{
 *   wpVersion: string,
 *   steps: list<StepDefinition>
 * }
 */

/**
 * @param Blueprint $blueprint
 * @return Blueprint
 */
function processBlueprint($blueprint) {
    // This is fine:
    echo $blueprint['wpVersion'];

    // This is a type error:
    echo $blueprint['no such key'];

    // These two lines do not trigger a type error, but
    // they cause the "return" statement to trigger one.
    // It would be cool if we could prevent those assignments
    // entirely.
    $blueprint['new_key'] = true;
    $blueprint['steps'][] = 'a string';
	return $blueprint;
}

/** @var Blueprint $blueprint */
processBlueprint([
    "wpVersion" => "6.4",
    "steps" => [
        [
            "step" => "installPlugin",
            "slug" => "gutenberg"
        ],
        [
            "step" => "rm",
            "path" => "file"
        ],
		// Uncomment this to get a type error
        // ["step" => "no such step"]
    ]
]);
```

**Comment by adamziel** [comment]
It seems like Psalm only gives strong guarantees when it deals with actual class instances:

```php
<?php

class Blueprint
{
    /** @var string */
    public $wpVersion;

    /** @var list<StepDefinition> */
    public $steps = [];
    
    public function __construct(string $wpVersion) {
     	$this->wpVersion = $wpVersion;   
    }
}

interface StepDefinition {}
class RmStep implements StepDefinition {
    /** @var "rm" */
    public $step = 'rm';
    /** @var string */
    public $path = 'rm';
}
class InstallPluginStep implements StepDefinition {
    /** @var "installPlugin" */
    public $step = 'installPlugin';
    /** @var string */
    public $slug = '';
}
class NotAStep {}

/**
 * @param Blueprint $blueprint
 * @return Blueprint
 */
function processBlueprint($blueprint) {
    // Fine:
    $blueprint->steps[] = new RmStep();
    $blueprint->steps[] = new InstallPluginStep();
    
    // Type errors:
    $blueprint->steps[] = new NotAStep();
    $blueprint->steps[] = (object)[
        'step' => 'rm'
    ];
    array_push($blueprint->steps, new NotAStep());
    
    // This is fine:
    echo $blueprint->wpVersion;

    // This is a type error:
    echo $blueprint->noSuchKey;

    // Type error
    $blueprint->new_key = true;
    return $blueprint;
}
```

Which means we'd still need to map raw objects into class instances, but we wouldn't need additional validation for the models – the type system would handle that for us. That isn't such a bad trade-off. Perhaps the `swaggest` code generator could be updated to generate psalm-annotated code?

**Comment by adamziel** [comment]
### PHPStan evaluation

[PHPStan](https://phpstan.org/) enables enforcing string typing rules without declaring a separate class for each data type:

```php
<?php declare(strict_types = 1);

/**
 * @phpstan-type RmStep object{step: "rm", path: string}
 * @phpstan-type InstallPluginStep object{step: "installPlugin", slug: string}
 * @phpstan-type Step RmStep|InstallPluginStep
 */
class Blueprint
{
	/** @var string */
	public $wpVersion = '6.4';
	
	/** @var Step[] */
	public $steps = [];
}

class BlueprintCompiler
{
	public function compile(Blueprint $blueprint): void
	{
		echo 'Hello, ' . $blueprint->steps[0]->step;
		$blueprint->wpVersion = "6.3";

		// Unfortunately no type error
		array_push($blueprint->steps, 'a');

		// Type error
		$blueprint->wpVersion = false;
		// Type error
		$blueprint->noSuchProperty = 123;

		// Type error
		$blueprint->steps[] = 'a';
		$blueprint->steps[] = (object)[
			'step' => 'rm',
			'path' => '/file'
		];

		// Type error
		$blueprint->steps[] = (object)[
			'step' => 'abc',
			'path' => '/file'
		];
	}
}
```

However, I don't see a way to instantiate a variable of a specific type without declaring a class of that type. E.g. this seems to be legal:

```php
/** @var InstallPluginStep $step */
$step = ['a'];
```

Although this is not:

```php
$step = 2;
/** @var InstallPluginStep $step */
$step['x'] = ['a']; // Cannot access offset 'x' on object{step: 'installPlugin', slug: string}.
```

**Comment by adamziel** [comment]
We have a way forward using code generation! In https://github.com/WordPress/blueprints/pull/19 I managed to assemble a model generation and mapping pipeline using [Nette PHP generator](https://github.com/nette/php-generator) and:

* Bits from the [janephp](https://github.com/janephp/janephp/) library to translate JSON schema into PHP models
* https://jsonmapper.net/ to translate parsed JSON object into model instances
* Validation isn't there yet, but it can be easily added via the https://opis.io/ library

This concludes the technically challenging part of this issue  🎉  Now it's all about defining the desired API shape. Adding a constructor, setters, or a static `create()` method is trivial at this point.

Example generated model:

```php
class InstallPluginStep implements StepDefinitionInterface
{
	public const DISCRIMINATOR = 'installPlugin';

	/** @var Progress */
	public $progress;

	/** @var bool */
	public $continueOnError;

	/**
	 * The step identifier.
	 * @var string
	 */
	public $step = 'installPlugin';

	/** @var string|ResourceDefinitionInterface */
	public $pluginZipFile;

	/** @var InstallPluginOptions */
	public $options;
}
```

**Comment by adamziel** [comment]
Blueprint validation, parsing, mapping, and a builder class just landed in PR #23. I wrote a comment to explain [how it fits the original design goals](https://github.com/WordPress/blueprints/pull/23#issuecomment-1961084826). This issue is now resolved 🎉
